rust/wasmtime_bindings/src/memory_creator.rs

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+use anyhow::{anyhow, Result};
+use std::{cmp, ptr, u32};
+use wasmtime::LinearMemory;
+
+const WASM_PAGE_SIZE: usize = 64 * 1024;
+
+extern "C" {
+    fn aligned_alloc(align: usize, size: usize) -> *mut u8;
+    fn free(ptr: *mut u8);
+}
+
+pub struct ScyllaLinearMemory {
+    ptr: *mut u8,
+    size: usize,
+    maximum_size: Option<usize>,
+}
+
+// The entire ScyllaLinearMemory is used only in a single thread,
+// because we're not sharing it between seastar shards
+unsafe impl Send for ScyllaLinearMemory {}
+unsafe impl Sync for ScyllaLinearMemory {}
+
+impl Drop for ScyllaLinearMemory {
+    fn drop(&mut self) {
+        // previously allocated or reset to nullptr in grow_to()
+        unsafe { free(self.ptr) };
+    }
+}
+
+unsafe impl LinearMemory for ScyllaLinearMemory {
+    fn byte_size(&self) -> usize {
+        self.size
+    }
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.maximum_size
+    }
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        let new_size_maybe_extra_page = new_size.checked_add(WASM_PAGE_SIZE - 1).ok_or(anyhow!(
+            "memory grow failed: new size {} is too large",
+            new_size
+        ))?;
+        let new_size_aligned = new_size_maybe_extra_page & !(WASM_PAGE_SIZE - 1);
+        if new_size_aligned == self.size {
+            return Ok(());
+        }
+        let max_size = self.maximum_size.unwrap_or(u32::MAX as usize + 1);
+        if new_size_aligned > max_size {
+            return Err(anyhow!(
+                "memory grow failed: new size {} exceeds maximum size {}",
+                new_size_aligned,
+                max_size
+            ));
+        }
+        let new_ptr: *mut u8;
+        if new_size_aligned == 0 {
+            new_ptr = ptr::null_mut()
+        } else {
+            new_ptr = unsafe { aligned_alloc(WASM_PAGE_SIZE, new_size_aligned) };
+            if new_ptr.is_null() {
+                return Err(anyhow!("Failed to grow WASM memory: allocation error"));
+            }
+        }
+        let copy_size = cmp::min(self.size, new_size_aligned);
+        if copy_size > 0 && !new_ptr.is_null() && !self.ptr.is_null() {
```

Comment:
Nit: `copy_size > 0` implies `!new_ptr.is_null() && !self.ptr.is_null()`, right? I don't think the null checks are required.

Reply:
Looks like it, I replaced them with a comment so that no one has doubts about it